### ISSUE TITLE
Activity functions fire multiple times for same input and don't all complete

### ISSUE THREAD

**Issue by mattschoutends** [Azure/azure-functions-durable-extension#1461]
### Description
Occasionally multiple instances of an activity function start at the same time.  And occasionally activity functions stop running without returning or throwing an exception.

Our function app loads large amounts of data from a file and stores it to a database.  We separate the records into batches and fan out.  Because of our database writes, our activity functions are not idempotent (though we attempted to simulate idempotency with one workaround.)

### Expected behavior
We expected that only one instance of an activity function would be started at the same time (i.e. if a second instance is created, it would not be at almost exactly the same time).

We also expected that if an activity function is started, it would run to completion.  This is the assumption that we based at least one of our workarounds on.

### Actual behavior
Original actual behavior:  we observed duplicate records being added to the database (or exceptions when unique constraints were violated).  This is what alerted us to the duplicate activity functions.

We added code to use table storage to ensure that only one activity function would add records (the "primary instance" of the activity function for the batch).  Duplicate instances of that activity function for the batch would return immediately.  In this case, we observed that not all the records would be added to the database.  We assumed that the framework would stop still-running instances with the same input when another instance completed.  The primary instance took longer to complete, so it would be terminated.

Then we added code so duplicate instances of activity functions for a batch would wait for the primary instance to complete (i.e. simulating idempotency, sort of).  In this case, we observed several instances where the primary instance never completed, but the primary instance also did not throw any exceptions.  The duplicate instance eventually timed out and threw an exception.

Our standard test splits the client data (where we see the issues) into 40 batches.  For the instance we reference below, we saw 45 instances of `SaveClients` executed.

### Attempted workarounds

- Upgraded to 2.3.0 (seemed somewhat better)
- Added a guard to limit to one writing instance per batch
- Caused duplicate instances to wait on the primary instance to complete

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: 2.3.0 (also observed with 2.2.2, 2.3.0 seems somewhat better)
- **Azure Functions runtime version (1.0 or 2.0)**: 2.0
- **Programming language used**: C#

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**:  2020-09-03 8:20-8:30 US Central time
- **Function App name**:  GlobalDataProd
- **Function name(s)**:  LoadMarketJson -> OrchestrateDataLoad -> SaveClients
- **Azure region**: CentralUS
- **Orchestration instance ID(s)**:  c2a41d9d35f14bccb3b806e0a99da556
- **Azure storage account name**:  globaldatafunctionprod

(We have other instances of failures if needed.)

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by cgillum** [comment]
Thanks @mattschoutends for these details. I found your orchestration instance in our internal telemetry and can confirm that it did run into some really odd behavior. I'll continue to take a look and see if I can determine the root cause and a possible fix/workaround.

**Comment by cgillum** [comment]
Hey @mattschoutends, so I looked at this a bit more and what I'm seeing is pretty confusing and counter-intuitive. Question for you: how are you determining that two functions are starting at the same time?

A couple things that I'm seeing:

1. In the Durable Task Framework layer, I do see what appears to be the same activity function starting twice concurrently. However, all traces associated with these executions are perfectly duplicated, including timestamps and message IDs, which seems impossible. From my end, it looks like a _duplicate logs issue_ on one of the app instances. Every log message on this instance is duplicated. However, if I cross-reference with telemetry that tracks raw function invocations, I don't see this duplication. The other app instances don't seem to have this problem at all.

2. I noticed that the instance that had all these duplicate executions seemed to have died mid-execution. This resulted in some of your activity functions being failed-over to another, healthy VM instance. In those cases, there were valid "second" executions of some functions, but they happened after a 5+ minute delay.

You mentioned that you were seeing two activity functions starting in the exact same time. Do you have a way of knowing whether there were _actually_ duplicate functions starting at the exact same time?

I haven't yet looked to see whether there were instances that didn't complete at all. I want to try to make sure I figure the duplication problem out first.

**Comment by mattschoutends** [comment]
Hi @cgillum, we have done several things over the past few days to try to understand what is going on.  In attempting to troubleshoot we've lost some history on our dev resource group, so I'm now realizing I can't go back and confirm what we were seeing.  It was a combination of function host traces (which it sounds like may be suspect) and application logs.

However, that said, in the information we still have, I can't find any evidence of two activity functions spinning up at the exact same time.  Is that something that could have been fixed by the update from 2.2.2 to 2.3.0 (thinking #1445)?

With the application-level data I have now, I'm seeing behavior consistent with what you noticed in item 2, a delay of at least a couple of minutes from the first execution to the second.  I'm basing that on the initial timestamp in our guard table (set by the first execution) and the timestamp of application log emitted when the second execution sees the guard table entry, correlated by the GUID assigned to the input batch.

I will double-check with the rest of my team to make sure I'm not forgetting anything and will run a few tests in dev to see if I can get the issue to reproduce.

> the instance that had all these duplicate executions seemed to have died mid-execution. This resulted in some of your activity functions being failed-over to another, healthy VM instance. In those cases, there were valid "second" executions of some functions, but they happened after a 5+ minute delay.

This is really interesting to me.  Is there a way for the application (orchestrator or activity function) to know when VM instances die?  Or when an activity function execution is abandoned/restarted due to a dead VM?  One reason I ask is that we've seen instance counts drop to zero in Live Metrics semi-regularly, or seen instances disappear.  I've mostly chalked that up to flakiness in Live Metrics behavior, but now I'm wondering if there's something else going on.

I don't want to sidetrack if this is not relevant to the issue you're chasing down, but seems like it could be related, at least in the sense of building applications that are reliable even when there might be retries.

**Comment by mattschoutends** [comment]
After comparing notes, we can't recall any instances of exact-same-time execution in the past few days, and definitely not since updating to 2.3.0.  It may have happened earlier or I may have misremembered when writing up the issue.  Does taking exact-same-time instances off the table help?

**Comment by rdmagm062699** [comment]
@cgillum Wondering if there has been any progress made on this issue?

**Comment by cgillum** [comment]
@mattschoutends sorry for the delayed response. I'm happy to hear you're not seeing more exact-same-time executions. Let me know if that changes. Let me try to answer your questions though:

> Is that something that could have been fixed by the update from 2.2.2 to 2.3.0 (thinking #1445)?

The fix you referenced would certainly help reduce the chance of duplication _generally_. However, it's not yet enabled by default so in theory it would only help you if you explicitly enabled it. However, the exact-same-time execution issue shouldn't be possible even in earlier versions, so I'm not really confident that I understand what was happening. To enable the new partition strategy that helps reduce duplication generally, you can set the `useLegacyPartitionManagement` setting to `false` in host.json (documented [here](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-bindings#durable-functions-2-0-host-json)).

> Is there a way for the application (orchestrator or activity function) to know when VM instances die? Or when an activity function execution is abandoned/restarted due to a dead VM?

It's possible for an activity function to know when the host is being shutdown by adding a `CancellationToken` to the function signature and observing the cancellation token getting signaled. However, in the case that I was looking at it seems the instance and/or VM died for unexpected reasons. There's unfortunately nothing we can do for those cases, but they should be rare. If you're looking for a way to know whether an activity function is undergoing a repeat execution due to such failures. we don't make this information available currently, but we could consider this for a new feature.

> One reason I ask is that we've seen instance counts drop to zero in Live Metrics semi-regularly, or seen instances disappear.

I assume instance counts is about Function host instances? If so, dropping to zero is expected if your app goes idle for ~20 minutes in the Consumption plan. In Elastic Premium, I think the app has to be idle for an hour, though I know the team was planning on increasing the idle timeout to something even larger, like 24 hours or more.

**Comment by mattschoutends** [comment]
@cgillum Thanks for getting back to us!  We have a few more days of experimentation and confusion.  Some things we've tried:

- Setting `useLegacyPartitionManagement`to `true` - did not seem to fix the issue.  There's a chance it may have reduced the number of duplications, but didn't remove them entirely.
- Deleting the task hub - seemed to help for the first execution or two afterwards, but not a long-term fix
- Setting `extendedSessionsEnabled` to `true` and setting the timeout to 2-3 minutes - seemed to significantly reduce the number of duplications.  I am guessing increasing the timeout further might "solve" the problem, but it seems like it's a band-aid over a deeper issue if it does.
- Refactoring to speed up our activity functions (more fan-outs and fan-ins) - seems to help, but only to the extent that longer-running functions seem more prone to vanishing into the ether.
    
>> One reason I ask is that we've seen instance counts drop to zero in Live Metrics semi-regularly, or seen instances disappear.

>I assume instance counts is about Function host instances? If so, dropping to zero is expected if your app goes idle for ~20 minutes in the Consumption plan. In Elastic Premium, I think the app has to be idle for an hour, though I know the team was planning on increasing the idle timeout to something even larger, like 24 hours or more.

I'm not sure if it's function host instances or VM/servers.  In Live Metrics, at the top of the screen it shows "# server online".  That's what I was referring to.  We regularly see an orchestration start, do some work, and then see that number drop to zero.  Then a few seconds or a minute later, it jumps to 1 or 2.  And sometimes we see a handful of servers online during a fan-out, then see the number drop, then climb back up.  This doesn't seem to quite line up with the ~20 minute idle--it's much shorter than that!

I'll take a look at cancellation tokens to see if they might help us design around the behavior we're seeing.

That reminds me--in one particularly bad test today, it seemed like the orchestrator rewound a step or two further than it should have.  The order of operations toward the end of the orchestration is like so:
1. Wait for all `SaveClients` activity functions to finish
1. Call a method named `SaveRevenueEvents` that fans out batches of work to activity functions (also named `SaveRevenueEvents`, assigning a GUID to each batch (generated using `context.NewGuid()`).  The method returns an `IEnumerable<Task>`.
1. Wait for all `SaveRevenueEvents` activity functions to complete.
We expected `n` instances of `SaveRevenueEvents`.  We saw `2n` (or maybe a handful more).  It looked like at some point, the orchestrator had been `await`ing (on step 4), but rewound and re-executed the `SaveRevenueEvents` method (step 2).  This was pretty unexpected.  I mention this because it seems related to replays and keeping track of state, but if it's not relevant, please disregard!  (Invocation ID was f01d4684-e023-4d9f-af8f-8db7b37bf8b1 in GlobalDataDev.)

We have continued to see the effect of this in both our dev and prod environments.  If there's anything we can do to help investigate, just say so; we'd be glad to help!

**Comment by cgillum** [comment]
> I'm not sure if it's function host instances or VM/servers. In Live Metrics, at the top of the screen it shows "# server online". That's what I was referring to. We regularly see an orchestration start, do some work, and then see that number drop to zero. Then a few seconds or a minute later, it jumps to 1 or 2. And sometimes we see a handful of servers online during a fan-out, then see the number drop, then climb back up.

There could be a couple explanations for this: 1) App Insights might not report it if it doesn't see any activity in less time than 20 minutes, and 2) I think when the platform scales-in, it might more aggressively take down the process. In that case, my 20 minute explanation is actually a bit misleading.

> in one particularly bad test today, it seemed like the orchestrator rewound a step or two further than it should have.

If an orchestration is scheduling the same execution multiple times, then it could be a problem checkpointing state. Looking at the telemetry for this orchestration (30630336df59499ab563e8a3ee646535), I did notice that it was running into split-brain issues. I checked and it seems like you did not actually apply the `useLegacyPartitionManagement` workaround I suggested (our telemetry reports that this value is still set to `true`, though I can't tell if that's because you explicitly set it to this or if you just left it as its default value). Can you please check that again?

**Comment by mattschoutends** [comment]
> If an orchestration is scheduling the same execution multiple times, then it could be a problem checkpointing state. Looking at the telemetry for this orchestration (30630336df59499ab563e8a3ee646535), I did notice that it was running into split-brain issues. I checked and it seems like you did not actually apply the useLegacyPartitionManagement workaround I suggested (our telemetry reports that this value is still set to true, though I can't tell if that's because you explicitly set it to this or if you just left it as its default value). Can you please check that again?

I just looked back at our history.  We did have `useLegacyPartitionManagement: false` when that orchestration ran.  We've since done other experiments and currently have that removed from our `host.json`.  

We're currently experimenting with the Elastic Premium plan to see if that improves things (it seems to be much more stable, but much more expensive)--we removed the `useLegacyPartitionManagement` entry this morning when we tried EP1.

Would it help for us to go back to the Consumption plan and put the `useLegacyPartitionManagement` workaround back in place?

**Comment by cgillum** [comment]
> I just looked back at our history. We did have useLegacyPartitionManagement: false when that orchestration ran. We've since done other experiments and currently have that removed from our host.json.

I checked again and for every VM that ran this orchestration, it reported that `useLegacyPartitionManagement` was set to `true`, so either I'm looking at the wrong instance ID or it might not have been set properly. Can you share your host.json file contents? For your workload, you should definitely be setting this to `false`. We're going to be changing the default to `false` in the next release.

**Comment by mattschoutends** [comment]
`host.json` as it was for that run:
```json
{
    "version": "2.0",
    "functionTimeout": "00:10:00",
    "extensions": {
        "durableTask": {
            "hubName": "%HUB_NAME%",
            "tracing": {
                "traceReplayEvents": true
            },
            "storageProvider": {
                "useLegacyPartitionManagement": false
            }
        }
    },
    "logging": {
        "applicationInsights": {
            "samplingSettings": {
              "isEnabled": false
            },
            "logLevel": {
                "DurableTask.AzureStorage": "Debug",
                "DurableTask.Core": "Debug"
              }
        }
    }
}
```

We're rolling back to Consumption in dev, have double-checked that `useLegacyPartitionManagement` is set to false.  Our first orchestration (40dc13f9ebd54c43bd477aceb3ab13c1) fired duplicate activity functions.

Everything was fine until our `SaveRevenueItems` activity functions started.  We expected ~40 instances.  We saw 6 start fairly normally.  3 of those completed.  It seemed like the whole app stalled.  About five minutes later, everything "woke up" again and fired the whole set of activity functions, including duplicates of the 6 that had already fired.

On our side, I'm going to try slightly changing the way we prepare and fire our activity functions.  Right now it's structured kind of like this:
```csharp
var batch = new List<T>();
foreach (var item in bigOldListOfItems) {
     batch.Add(item);
     if (IsBigEnough(batch)) {  // getting to "big enough" might take a while
          context.CallActivityAsync("SaveRevenueItems", batch);
          batch.Clear();
     }
}
```
I suspect it might help to build the batches first, then fire all the activity functions right away:
```csharp
var batches = new List<List<T>>();
var batch = new List<T>();
foreach (var item in bigOldListOfItems) {
     batch.Add(item);
     if (IsBigEnough(batch)) {  // getting to "big enough" might take a while
          batches.Add(batch);
          batch = new List<T>();
     }
}
batches.Select(x => context.CallActivityAsync("SaveRevenueItems", x));
```

It's not clear to me why the orchestrator would stop in the middle of a method, what might cause it to do that, or if that's documented/expected behavior.

**Comment by mattschoutends** [comment]
Quick update, and thanks for your ongoing investigation, @cgillum.

The orchestration mentioned above, 40dc13f9ebd54c43bd477aceb3ab13c1, exhibited some very strange behavior after activity functions began throwing exceptions.  It seems to have continued rewinding and creating new activity functions.  It was still running in a loop of creating new functions, throwing exceptions, and rewinding 4-5 hours later.  I happened to notice and manually terminated it--it took quite a few attempts with `func durable terminate`, but it finally stopped.  Just one durable instance in a brand new hub.

**Comment by cgillum** [comment]
@mattschoutends I took a look at 40dc13f9ebd54c43bd477aceb3ab13c1 and I can confirm that `useLegacyPartitionManagement` was correctly set to `false`, so that rules out partition management as the cause of the duplication.

I took a look at one particular `SaveRevenueItems` activity duplication to see what might be going on. What I observed is that the functions host is being shut down while the orchestration is in the middle of a lengthy checkpoint operation. It looks like it took over 60 seconds (which is very unusual) to checkpoint a particular step and the shutdown happened partway through. Checkpoints in Durable Functions are not atomic, so the result of the partial checkpoint is that some activity functions got scheduled mid-way through but the orchestration itself was reverted back to the last known good state, resulting in them getting scheduled a second time.

The big question for me is why did the checkpoint take so long. I don't have good insights into that, though high CPU or other resource contention might explain it. I did observe a couple out-of-memory exceptions in your app, but I'm not sure to what extent those could have impacted your test run.

I'm not immediately sure how to remedy this situation. The simple thing you can do is deploy your app to a dedicated App Service Plan to ensure that instances of your host are never scaled-in and shut down abruptly. Sorry for the inconvenience...

**Comment by mattschoutends** [comment]
@cgillum Believe it or not, your update was really helpful.  Our application keeps a lot of data in memory and parcels it out to activity functions.  The hypothesis we developed based on your update was that the framework was taking a long time to checkpoint (and start activity functions, and probably return from activity functions) because of the amount of data being passed around (serialization/transmission/deserialization, memory allocation, possibly even matching up a function result with the input).  

As an experiment, we changed the way we handled data.  Originally the orchestrator did something like this:
```csharp
var lotsOfData = await LoadUpLotsOfData(filename);
var topLevelId = await SaveTopLevelOfHierarchy(lotsOfData.topLevelThing);
var nextLevelIds = await SaveNextLevelOfHierarchy(topLevelId, lotsOfData.nextThing);
var thirdLevelIds = await SaveThirdLevelOfHierarchy(topLevelId, nextLevelIds, lotsOfData); // creates batches, fans out, fans in
var fourthLevelIds = await SaveFourthLevelOfHierarchy(topLevelId, thirdLevelIds, lotsOfData); // batches, fan-out/fan-in
// etc
```

We revised it to do something more like this:
```csharp
var lotsOfData = await LoadUpLotsOfData(filename);  // Creates batches in blobs and returns blob names with all the other data, so actually returns more than original
var topLevelId = await SaveTopLevelOfHierarchy(lotsOfData.topLevelThing);  // unchanged
var nextLevelIds = await SaveNextLevelOfHierarchy(topLevelId, lotsOfData.nextThing);  // unchanged
var thirdLevelIds = await SaveThirdLevelOfHierarchy(topLevelId, nextLevelIds, lotsOfData.batchBlobNames);  // fan out activity functions, but input is just the IDs and a filename - many, many fewer bytes
var fourthLevelIds = await SaveFourthLevelOfHierarchy(topLevelId, thirdLevelIds, lotsOfData.batchBlobNames); // fan out activity functions, but input is just the IDs and a filename - many, many fewer bytes
// etc
```

On Elastic Premium (EP1) everything became much smoother and now just works.  We ran into some issues with Consumption (long pauses between orchestrator steps, mostly, but also a few times when server counts would drop to zero), so we decided to stick to EP1 for now.

Anyway, I expect that the checkpointing taking so long may have had to do with large inputs to the activity functions and our application generally hitting the storage account pretty hard.  From what I understand, durable functions is already a pretty significant load on a storage account, let alone when the applications are moving around big chunks of data.  Our average and max latency numbers on the storage account are much lower now than they were before this change.

Not sure if there's still an issue worth looking into with large inputs to activity functions, or with handling increased latency with storage accounts as load goes up.  

Thanks for your help on this, and for insight into things that we didn't or couldn't see!

**Comment by ConnorMcMahon** [comment]
@mattschoutends,

Large messages definitely can account for longer to checkpoint. If we are sending large queue messages and/or large table rows, we are forced to write messages to blob storage, which is inherently a more expensive operation. It sounds like given the improved perf you saw when chunking, this could have been a major part of your latency.

If you find your team experimenting with Consumption plan again and run into more issues there, feel free to open a ticket with timestamps and slow instance ids and we can investigate that, as we are trying to make sure that we can get acceptable performance across dedicated, consumption, and premium plans, so any cases where consumption isn't working as desired is something we will want to pursue.
